design/metal3host-controller.md

```diff
@@ -0,0 +1,314 @@
+# Baremetal host single instance and multi-instance operator
+
+## Introduction
+
+Currently, the metal3 system is to set up a Kubernetes cluster on bare-metal hosts which is tightly bound with cluster-api and Kubernetes. We target to allocate bare-metal hosts that just pre-installing the OS, or pre-installing the applications as required.
+
+## Motivation
+
+To meet the scenery that only requires a pure bare-metal host or with a simple application installed.
+
+## Goals
+
+- Allocate one pure bare-metal host
+- Allocate multiple bare-metal hosts
+- Allocate bare-metal host with pre-installed applications
+- Specify the different configurations for different bare-metal host
+
+## Non-goals
+
+- Allocate different types of bare-metal hosts at one time
+
+## Proposal
+
+### Architecture
+
+The operators include the metal3Host operator, metal3HostDeployment operator, and data operator.
+![modules](images/metal3hostModules.png)
+
+The bare-metal host controller and the ironic are not newly developed.
+
+Metal3Host is a virtual instance. Users could define it to require a bare-metal host. In the definition, the user could specify the image to install, could set the selector to choose a specific bare-metal host, and could define the boot data to automatically start their own applications.
```

Comment:
Thanks for your review.
The difference between metal3Host and baremetalHost is that metal3Host is a virtual host. When the user requires a host, a metal3Host instance will be created. The metal3Host instance has some features the user defines, such as the operating system, the user name/password, the bootup application, etc. After the instance is created, it will choose one suitable baremetalHost.

About the name, I'm not clear which is best.

Thanks & Regards,
Wei

Reply:
Another side, if one baremetalHost is down, the metal3Host could choose another baremetalHost. We also could label the baremetalHost, so users could define their requirements such as they need a strong CPU host. Then the metal3Host could choose a host by the label. The metal3Host represents the users' requirement, the baremetal3Host represents the physical machine. I think it's more flexible.

Reply:
I see so the host is handled like a workload, like a pod.

Reply:
Okay, I can agree with a new object, but I cannot agree with calling it Metal3Host. BareMetalHost is also a Metal3 host technically. Could be BareMetalInstance, ProvisionedBareMetal, BareMetalProvisioning... Having a Metal3 prefix is not terribly helpful, definitely does not solve the potential confusion.

Reply:
https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit has some prior art with regards to naming.